### PR TITLE
Implement semantic version based intelligent routing for APIs

### PR DESCRIPTION
**Pull request by chathuranga95**
### Purpose

This PR adds the reverted changes from: https://github.com/wso2/product-microgateway/pull/3376 and https://github.com/wso2/product-microgateway/pull/3407 to the product.

Also this PR fixes the issue where the resource path substitution string gets converted to a regex

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/internal/oasparser/envoyconf/routes_with_clusters.go [hunk 1]
```diff
@@ -1353,12 +1360,16 @@ func generatePathRegexSegment(resourcePath string) string {
 	return resourceRegex
 }
 
+func generatePathRegexSegmentForSubstitutionString(resourcePath string) string {
```
Comment (VirajSalaka) [inline adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1363]:
```suggestion
// generatePathRegexSegmentForSubstitutionString generates the regex for substitution string. Here the dots should not be transformed to regex compared to generating router matching regex but the path params are. 
func generatePathRegexSegmentForSubstitutionString(resourcePath string) string {
```

Comment (chathuranga95) [reply adapter/internal/oasparser/envoyconf/routes_with_clusters.go:1366]: Added the function description with: https://github.com/wso2/product-microgateway/pull/3415/commits/084b4812aa3792bb909882f644265650c91f4358

**Review by VirajSalaka — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[failed] Controlplane cluster : dev-deployment-v2 : 20230807.14

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : dev-deployment-v2 : 20230807.14

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20230807.15

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : dev-deployment-v2 : 20230807.15

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20230807.3

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : stage-deployment-v2 : 20230807.3

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : stage-deployment-v2 : 20230807.3

**Comment by choreo-cicd** [comment]
[failed] Controlplane cluster : prod-deployment-v2 : 20230807.4

**Comment by choreo-cicd** [comment]
[failed] Dataplane(EastUS) cluster : prod-deployment-v2 : 20230807.4

**Comment by choreo-cicd** [comment]
[failed] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20230807.4

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20230807.4

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : prod-deployment-v2 : 20230807.4

**Pull request merged**